sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx

```diff
@@ -0,0 +1,245 @@
+#include "StubMergeAlgorithms.h"
+
+geo::Point_t sbn::GetLocation(const spacecharge::SpaceCharge *sce, geo::Point_t loc_w, unsigned TPC, float xsign) {
+  if (sce && sce->EnableCalSpatialSCE()) {
+    geo::Vector_t offset = sce->GetCalPosOffsets(loc_w, TPC);
+    loc_w.SetX(loc_w.X() + xsign * offset.X());
+    loc_w.SetY(loc_w.Y() + offset.Y());
+    loc_w.SetZ(loc_w.Z() + offset.Z());
+  }
+
+  return loc_w;
+}
+
+double sbn::GetEfield(const detinfo::DetectorPropertiesData &dprop, const spacecharge::SpaceCharge *sce, geo::Point_t loc, unsigned TPC, bool correct_loc_sce, float xsign) {
+
+  double EField = dprop.Efield();
+  if (sce && sce->EnableSimEfieldSCE()) {
+    if (correct_loc_sce) loc = sbn::GetLocation(sce, loc, TPC, xsign);
+
+    // Gets relative E field Distortions
+    geo::Vector_t EFieldOffsets = sce->GetEfieldOffsets(loc);
+    // Add 1 in X direction as this is the direction of the drift field
+    EFieldOffsets = EFieldOffsets + geo::Vector_t{1, 0, 0};
+    // Convert to Absolute E Field from relative
+    EFieldOffsets = EField * EFieldOffsets;
+    // We only care about the magnitude for recombination
+    EField = EFieldOffsets.r();
+  }
+
+  return EField;
+}
+
+geo::Point_t sbn::GetLocationAtWires(const spacecharge::SpaceCharge *sce, geo::Point_t loc, float xsign) { 
+  if (sce && sce->EnableCalSpatialSCE()) {
+    // fix negative sign in TPC 0
+    int corr = 1;
+    float xx = loc.X();
+    if (xx < 0) { corr = -1; }
+
+    // for some reason, one needs to flip the sign of the x-direction when correcting for field distortion
+    geo::Vector_t offset = sce->GetPosOffsets(loc);
+    loc.SetX(loc.X() + corr * xsign * offset.X());
+    loc.SetY(loc.Y() + offset.Y());
+    loc.SetZ(loc.Z() + offset.Z());
+  }
+
+  return loc;
+}
+
+double sbn::GetPitch(
```

Comment:
This is the same as the Pitch calculation that is done in the calorimetry. It needs to handle a few different cases (presence of the space charge or not; whether the absolute positions of the PFParticle are changed).

The key thing though is that the presence of space charge physically changes the pitch beyond the correction of the absolute positions on either end of the wire.

I guess there is a function in the geometry that one can "outsource" some of the math to. I don't think it is necessarily advantageous for me to use it. The math I have written in this function is not wrong, and I can see what it is doing by direct inspection. And since it's just a piece of math, I don't anticipate having to change it in the future.